File: esp32/acurite609.cpp
#include "acumonitor.h"

/**
 * Parsing && chunk-building for model-specific RF signals.
 */
Acurite609::Model::Model(std::vector<Acurite609::Device> devices) {
    this->devices = devices;
}

void Acurite609::Model::clear() {
    bitstream = 0;      // Will contain all bits received in a single bitstream
    bitstream_size = 0; // Size in bits of current bitstream
    last_rfs_type = ACURITE609_SIGNAL_INV;
    bitstream_open = false;
    // Only manually reset chunk status in parse_rf
}

int Acurite609::Model::get_rfs_type(uint8_t rfs, uint32_t duration) {
    /* Returns the type of RF signal received.

    :param int rfs: RF signal received; either 0 || 1
    :param int duration: signal duration, in microseconds
    :return: the signal type
    :rtype: int
    */
    if (rfs == 0) {
        if (duration < 1200)
            return ACURITE609_SIGNAL_OFF;
    }
    else if (rfs == 1) {
        if (duration < 300)
            return ACURITE609_SIGNAL_CHUNK_START;
        else if (duration >= 300 && duration < 1200)
            return ACURITE609_SIGNAL_BIT_0;
        else if (duration >= 1200 && duration < 3000)
            return ACURITE609_SIGNAL_BIT_1;
        else if (duration >= 8700 && duration < 9000)
            return ACURITE609_SIGNAL_BITSTREAM_START;
        else if (duration >= 10000 && duration < 20000)
            return ACURITE609_SIGNAL_BITSTREAM_END;
        else if (duration >= 20000 && duration < 40000)
            return ACURITE609_SIGNAL_CHUNK_END;
    }
    return ACURITE609_SIGNAL_INV;
}

bool Acurite609::Model::is_bit_signal(int rfs_type) {
    return rfs_type == ACURITE609_SIGNAL_BIT_0 || rfs_type == ACURITE609_SIGNAL_BIT_1;
}

void Acurite609::Model::open_bitstream() {
    bitstream_open = true;
    bitstream_size = 0;
    bitstream = 0;
}

void Acurite609::Model::close_bitstream() {
    bitstream_open = false;
    bitstream_size = 0;
    bitstream = 0;
}

void Acurite609::Model::open_chunk() {
    chunk_open = true;
    open_bitstream();
}

void Acurite609::Model::close_chunk() {
    chunk_open = false;
    close_bitstream();
}

uint64_t Acurite609::Model::parse_rf(uint32_t duration, uint8_t rfs) {
    /* Parse a single RF signal && update chunk/bitstreams.
     */
    uint64_t result = 0;
    int rfs_type = ACURITE609_SIGNAL_INV;
    rfs_type = get_rfs_type(rfs, duration);
    // Last signal must be ACURITE609_SIGNAL_OFF
    if (last_rfs_type == ACURITE609_SIGNAL_OFF && !chunk_open) {
        if (rfs_type == ACURITE609_SIGNAL_BITSTREAM_START)
            open_chunk();
    }
    else if (last_rfs_type == ACURITE609_SIGNAL_OFF && chunk_open) {
        if (rfs_type == ACURITE609_SIGNAL_BITSTREAM_START && !bitstream_open) {
            if (bitstream_size == ACURITE609_SIGNAL_BIT_LENGTH)
                result = bitstream;
            open_bitstream();
        }
        else if (rfs_type == ACURITE609_SIGNAL_BITSTREAM_END && bitstream_open) {
            if (bitstream_size == ACURITE609_SIGNAL_BIT_LENGTH)
                result = bitstream;
            close_bitstream();
        }
        else if (rfs_type == ACURITE609_SIGNAL_CHUNK_END) {
            last_rfs_type = rfs_type;
            if (bitstream_size == ACURITE609_SIGNAL_BIT_LENGTH)
                result = bitstream;
            close_chunk();
        }
        else if (is_bit_signal(rfs_type) && bitstream_open) {
            if (rfs_type == ACURITE609_SIGNAL_BIT_1 && bitstream_size < ACURITE609_SIGNAL_BIT_LENGTH)
                bitstream |= ((uint64_t)1L << (ACURITE609_SIGNAL_BIT_LENGTH - bitstream_size - 1));
            bitstream_size += 1;
            if (bitstream_size == ACURITE609_SIGNAL_BIT_LENGTH) {
                result = bitstream;
                close_bitstream();
            }
        }
    }
    last_rfs_type = rfs_type;

    // Done
    return result;
}

Acurite609::Device::Device(uint16_t device) {
    this->device = device;
    this->temperature = 0;
    this->humidity = 0;
    this->battery = 0;
    this->signature = 0;
}

Payload *Acurite609::Device::create_payload(uint8_t status) {
    Payload *payload = new Payload();
    payload->tag = TAG_TEMPMONITOR;
    payload->model = MODEL_ACURITE609;
    payload->device = device;
    payload->status = status;
    payload->battery = battery;
    payload->temperature = int16_t(temperature * 10);
    payload->humidity = int16_t(humidity * 10);
    return payload;
}

bool Acurite609::Device::validate_checksum(uint64_t bitstream) {
    uint8_t checksum = bitstream & 0xff;
    uint32_t calculated = (((bitstream >> 8) & 0xff) + 
            ((bitstream >> 16) & 0xff) +
            ((bitstream >> 24) & 0xff) + 
            ((bitstream >> 32))) & 0xff;
    if (checksum != calculated) {
        Serial.print("bad checksum: ");
        Serial.print(checksum, HEX);
    }
    return checksum == calculated;
}

/**
 * Validates the signature && checksum of the specified bitstream.
 *
 * @param bitstream bitstream of data to validate
 * @return true if bitstream is good, false if bad
 */
bool Acurite609::Device::validate_bitstream(uint64_t bitstream) {
    if (bitstream == 0)
        return false;
    Serial.print("[");
    Serial.print(bitstream, BIN);
    Serial.print("] ");
    uint16_t sig = bitstream >> 32;
    if (signature != 0 && signature != sig) {
        Serial.print("bad signature: ");
        Serial.println(sig, HEX);
        return false;
    }
    int cha = (bitstream >> 28) & 0x03;
    if (cha != ACURITE609_CHANNEL_ID) {
        Serial.print("bad channel: ");
        Serial.println(cha);
        return false;
    }
    if (!validate_checksum(bitstream)) {
        Serial.print(", signature: ");
        Serial.println(sig, HEX);
        return false;
    }
    uint8_t bat = (bitstream >> 30) & 0x03;
    float temp = (bitstream >> 15) & 0x1fff;
    if ((uint16_t)temp & 0x1000 == 0x1000)
        temp = -(0x2000 - temp);
    temp /= 20;
    float hum = (bitstream >> 8) & 0x7f;
    if (hum < 1 || hum > 99 || temp < -40 || temp > 70) {
        Serial.print("invalid data: ");
        Serial.print(temp);
        Serial.print("C ");
        Serial.print(hum);
        Serial.println("%");
        return false;
    }
    // Set the instance values
    if (signature == 0)
        signature = sig;
    battery = bat;
    humidity = hum;
    temperature = temp;
    Serial.print("outdoor: ");
    Serial.print(temp * 1.8 + 32);
    Serial.print("F ");
    Serial.print(hum);
    Serial.print("%, battery=");
    Serial.println(bat);
    return true;
}

File: rpi/acurite609.py
#!/usr/bin/python3 -u

"""
Parses 433MHz RF signals from the AcuRite 00609SBLA2 outdoor unit.
"""

import struct

SIGNAL_BIT_LENGTH = 40          # Length of bitstream in bits
SIGNAL_INV = -2
SIGNAL_OFF = -1
SIGNAL_BIT_0 = 0
SIGNAL_BIT_1 = 1
SIGNAL_BITSTREAM_START = 2
SIGNAL_BITSTREAM_END = 3
SIGNAL_CHUNK_START = 4
SIGNAL_CHUNK_END = 5
CHANNEL_ID = 2
MODEL_ACURITE523 = 1592
MODEL_ACURITE609 = 6585
TAG_TEMPMONITOR = 0x38073162

class Acurite609():
    class Model():
        """Parsing and chunk-building for model-specific RF signals.
        """
        def __init__(self, devices):
            self.devices = devices
            self.bitstream = 0       # Will contain all bits received in a single bitstream
            self.bitstream_size = 0  # Size in bits of current bitstream
            self.last_rfs_type = SIGNAL_INV
            self.bitstream_open = False
            self.chunk_open = False

        def clear(self):
            self.bitstream = 0
            self.bitstream_size = 0
            self.last_rfs_type = SIGNAL_INV
            self.bitstream_open = False
            self.chunk_open = False

        def get_rfs_type(self, rfs, duration):
            """Returns the type of RF signal received.

            :param int rfs: RF signal received; either 0 or 1
            :param int duration: signal duration, in microseconds
            :return: the signal type
            :rtype: int
            """
            if rfs == 0:
                if duration < 1200:
                    return SIGNAL_OFF
            elif rfs == 1:
                if duration < 300:
                    return SIGNAL_CHUNK_START
                elif duration >= 300 and duration < 1200:
                    return SIGNAL_BIT_0
                elif duration >= 1200 and duration < 3000:
                    return SIGNAL_BIT_1
                elif duration >= 8700 and duration < 9000:
                    return SIGNAL_BITSTREAM_START
                elif duration >= 10000 and duration < 20000:
                    return SIGNAL_BITSTREAM_END
                elif duration >= 20000 and duration < 40000:
                    return SIGNAL_CHUNK_END
            return SIGNAL_INV

        def is_bit_signal(self, rfs_type):
            return rfs_type == SIGNAL_BIT_0 or rfs_type == SIGNAL_BIT_1

        def open_bitstream(self):
            self.bitstream_open = True
            self.bitstream_size = 0
            self.bitstream = 0

        def close_bitstream(self):
            self.bitstream_open = False
            self.bitstream_size = 0
            self.bitstream = 0

        def open_chunk(self):
            self.chunk_open = True
            self.open_bitstream()

        def close_chunk(self):
            self.chunk_open = False
            self.close_bitstream()

        def parse_rf(self, duration, rfs):
            """Parse a single RF signal and update chunk/bitstreams.
            """
            result = 0
            rfs_type = SIGNAL_INV
            rfs_type = self.get_rfs_type(rfs, duration)
            # Last signal must be SIGNAL_OFF
            if self.last_rfs_type == SIGNAL_OFF and not self.chunk_open:
                if rfs_type == SIGNAL_BITSTREAM_START:
                    self.open_chunk()
            elif self.last_rfs_type == SIGNAL_OFF and self.chunk_open:
                if rfs_type == SIGNAL_BITSTREAM_START and not self.bitstream_open:
                    if self.bitstream_size == SIGNAL_BIT_LENGTH:
                        result = self.bitstream
                    self.open_bitstream()
                elif rfs_type == SIGNAL_BITSTREAM_END and self.bitstream_open:
                    if self.bitstream_size == SIGNAL_BIT_LENGTH:
                        result = self.bitstream
                    self.close_bitstream()
                elif rfs_type == SIGNAL_CHUNK_END:
                    self.last_rfs_type = rfs_type
                    if self.bitstream_size == SIGNAL_BIT_LENGTH:
                        result = self.bitstream
                    self.close_chunk()
                elif self.is_bit_signal(rfs_type) and self.bitstream_open:
                    if rfs_type == SIGNAL_BIT_1 and self.bitstream_size < SIGNAL_BIT_LENGTH:
                        self.bitstream |= (1 << (SIGNAL_BIT_LENGTH - self.bitstream_size - 1))
                    self.bitstream_size += 1
                    if self.bitstream_size == SIGNAL_BIT_LENGTH:
                        result = self.bitstream
                        self.close_bitstream()
            self.last_rfs_type = rfs_type
            # Done
            return result

    class Device():
        def __init__(self, device, verbosity=0):
            self.print_verbose = print if verbosity > 1 else lambda *a, **k: None
            self.print_debug = print if verbosity > 2 else lambda *a, **k: None
            self.device = device
            self.temperature = 0
            self.humidity = 0
            self.battery = 0
            self.signature = 0

        def create_payload(self, status):
            return struct.pack('<IHHbbhh', TAG_TEMPMONITOR, 
                    MODEL_ACURITE609, self.device, status, self.battery, 
                    int(self.temperature * 10), int(self.humidity * 10))

        def validate_checksum(self, bitstream):
            checksum = bitstream & 0xff
            calculated = (((bitstream >> 8) & 0xff) + 
                    ((bitstream >> 16) & 0xff) +
                    ((bitstream >> 24) & 0xff) + 
                    ((bitstream >> 32))) & 0xff
            if checksum != calculated:
                self.print_verbose(f'bad checksum: {calculated:#x}', end='')
            return checksum == calculated

        def validate_bitstream(self, bitstream):
            """Validates the signature and checksum of the specified bitstream.

            :param int bitstream: the bitstream of data to validate
            :return: True if bitstream is good, False if bad
            """
            if bitstream == 0:
                return False
            self.print_verbose(f'[{bitstream:040b}] ', end='')
            sig = bitstream >> 32
            if self.signature != 0 and self.signature != sig:
                self.print_verbose(f'bad signature: {sig:#x}')
                return False
            cha = (bitstream >> 28) & 0x03
            if cha != CHANNEL_ID:
                self.print_verbose(f'bad channel: {cha}')
                return False
            if not self.validate_checksum(bitstream):
                self.print_verbose(f', signature: {sig:#x}')
                return False
            bat= (bitstream >> 30) & 0x03
            temp = (bitstream >> 15) & 0x1fff
            if temp & 0x1000 == 0x1000:
                temp = -(0x2000 - temp)
            temp /= 20
            hum = (bitstream >> 8) & 0x7f
            if hum < 1 or hum > 99 or temp < -40 or temp > 70:
                self.print_verbose(f'invalid data: {temp}C {hum}%')
                return False
            # Set the instance values
            if self.signature == 0:
                self.signature = sig
            self.battery = bat
            self.humidity = hum
            self.temperature = temp
            temp = temp * 1.8 + 32
            self.print_verbose(f'outdoor: {temp:.1f}F {hum}%, battery={bat}')
            return True

